README.md
# RoutmouteDiscordBundle

## Manual Installation

Make sure Composer is installed globally, as explained in the
[installation chapter](https://getcomposer.org/doc/00-intro.md)
of the Composer documentation.

### Step 1: Create configuration file

Create configuration file `config/packages/routmoute_discord.yaml` and modify scopes if you want

```yaml
// config/packages/routmoute_discord.yaml

routmoute_discord:
    oauth:
        client_id: '%env(ROUTMOUTE_DISCORD_CLIENT_ID)%'
        client_secret: '%env(ROUTMOUTE_DISCORD_CLIENT_SECRET)%'
        scope:
            - identify
            - email
    api:
        bot_token: '%env(ROUTMOUTE_DISCORD_BOT_TOKEN)%'
```

### Step 2: Download the Bundle

Open a command console, enter your project directory and execute the
following command to download the latest stable version of this bundle:

```console
composer require routmoute/routmoute-discord-bundle
```

## Configuration

### Step 1: Create your Discord Application

- Go to <https://discord.com/developers/applications>
- Create a New Application
- Copy `CLIENT ID` and `CLIENT SECRET` for next step
- Go to `OAuth2` Tab
- Add Redirect `https://yourDomain.domain/receiveDiscord`
- Go to Bot Tab and copy `TOKEN` for next step

### Step 2: Create your env variables

Add this environments vars in your `.env` file.

```
ROUTMOUTE_DISCORD_CLIENT_ID=YourClientId
ROUTMOUTE_DISCORD_CLIENT_SECRET=YourClientSecret
ROUTMOUTE_DISCORD_BOT_TOKEN=YourBotToken
```

## Usage (for Symfony 5)

### Discord OAuth

#### Step 1: Create Controller in your App

Create Controller, for exemple `src/Controller/DiscordOAtuhController.php`

```php
// src/Controller/DiscordOAtuhController.php

<?php

namespace App\Controller;

use Routmoute\Bundle\RoutmouteDiscordBundle\Service\RoutmouteDiscordOAuthService;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\RedirectResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\Routing\Generator\UrlGeneratorInterface;

class DiscordOAuthController extends AbstractController
{
    /**
    * @Route("/connectToDiscord", name="routmoute_discord_redirect", methods="GET")
    */
    public function redirectToDiscord(RoutmouteDiscordOAuthService $oAuthService, UrlGeneratorInterface $urlGenerator): RedirectResponse
    {
        $redirectUrl = $urlGenerator->generate('routmoute_discord_receiver', [], UrlGeneratorInterface::ABSOLUTE_URL);
        return new RedirectResponse($oAuthService->getRedirectDiscordUrl($redirectUrl));
    }

    /**
    * @Route("/receiveDiscord", name="routmoute_discord_receiver", methods="GET")
    */
    public function receiveFromDiscordAuthorize(Request $request, RoutmouteDiscordOAuthService $oAuthService, UrlGeneratorInterface $urlGenerator): RedirectResponse
    {
        $redirectUrl = $urlGenerator->generate('routmoute_discord_receiver', [], UrlGeneratorInterface::ABSOLUTE_URL);
        $userData = $oAuthService->getUserData($request, $redirectUrl);

        // TODO: Process userData and change path_to_redirect

        return $this->redirectToRoute('path_to_redirect');
    }
}
```

#### Step 2: Create your redirect button

Create a button in your frontend that redirect to `routmoute_discord_redirect` path.

for example, in twig template:

```html
<a href="{{ path('routmoute_discord_redirect') }}">
    <button type="button">Link my account with discord</button>
</a>
```

### Discord API (Bot)

Example usage in Controller:

```php
<?php
namespace App\Controller;

use Routmoute\Bundle\RoutmouteDiscordBundle\Service\RoutmouteDiscordApiService;

class MyController extends AbstractController
{
    public function index(RoutmouteDiscordApiService $discordAPI)
    {
        $discordId = 'theUserDiscordId';

        $userInfos = $discordAPI->getUserFromDiscordId($discordId);

        $userInfos["id"];
        $userInfos["username"];
        $userInfos["discriminator"];
        $userInfos["avatar"];
    }
}
```

<https://discord.com/developers/docs/resources/user>

## Parameters

### `client_id`

_Required_
The `CLIENT ID` provided by discord

### `client_secret`

_Required_
The `CLIENT SECRET` provided by discord

### `scope`

_Required_
The Discord API scopes - <https://discord.com/developers/docs/topics/oauth2#shared-resources>

- `identify` - discordId, avatar, username, discriminator
- `email` - email
- ...

### `bot_token`

_Required_
The Bot `TOKEN` provided by discord


CHANGELOG.md
# Changelog

## NEXT
 * Add Exceptions

## v1.1.0
 * [#1](https://github.com/routmoute/routmoute-discord-bundle/pull/1) Add DiscordApiService

## v1.0.1
 * Fix for Symfony 4.4

## v1.0.0
* Initial release
* obtains userDatas after obtaining user consent.


tests/RoutmouteDiscordTestKernel.php
<?php

namespace Routmoute\Bundle\RoutmouteDiscordBundle\Tests;

use Symfony\Component\HttpKernel\Kernel;
use Symfony\Component\Config\Loader\LoaderInterface;

class RoutmouteDiscordTestKernel extends Kernel
{
    public function registerBundles()
    {
        return [
            new \Symfony\Bundle\FrameworkBundle\FrameworkBundle(),
            new \Routmoute\Bundle\RoutmouteDiscordBundle\RoutmouteDiscordBundle(),
        ];
    }

    public function registerContainerConfiguration(LoaderInterface $loader)
    {
        $loader->load(__DIR__.'/config/config.yaml');
    }
}


tests/Unit/Service/RoutmouteDiscordApiServiceTest.php
<?php

namespace Routmoute\Bundle\RoutmouteDiscordBundle\Tests\Unit\Service;

use PHPUnit\Framework\TestCase;
use Routmoute\Bundle\RoutmouteDiscordBundle\Service\RoutmouteDiscordApiService;
use Symfony\Component\HttpClient\MockHttpClient;
use Symfony\Component\HttpClient\Response\MockResponse;

class RoutmouteDiscordApiServiceTest extends TestCase
{
    private $botToken;
    private $discordId;
    private $guildId;

    public function testGetUserFromDiscordId()
    {
        $this->botToken = $this->generateRandomString();
        $this->discordId = $this->generateRandomString();

        $mockHttpClient = new MockHttpClient(function($method, $url, $options) {
            $this->assertEquals("GET", $method);
            $this->assertEquals('https://discord.com/api/users/' . $this->discordId, $url);
            $this->assertEquals($options['headers'][1], "Authorization: Bot " . $this->botToken);

            return new MockResponse(json_encode([]), ['headers' => ['Content-Type' => 'application/json']]);
        });

        $apiService = new RoutmouteDiscordApiService($this->botToken, $mockHttpClient);

        $apiService->getUserFromDiscordId($this->discordId);
    }

    public function testGetUserFromGuild()
    {
        $this->botToken = $this->generateRandomString();
        $this->discordId = $this->generateRandomString();
        $this->guildId = $this->generateRandomString();

        $mockHttpClient = new MockHttpClient(function($method, $url, $options) {
            $this->assertEquals("GET", $method);
            $this->assertEquals("https://discord.com/api/guilds/" . $this->guildId . "/members/" . $this->discordId, $url);
            $this->assertEquals($options['headers'][1], "Authorization: Bot " . $this->botToken);

            return new MockResponse(json_encode([]), ['headers' => ['Content-Type' => 'application/json']]);
        });

        $apiService = new RoutmouteDiscordApiService($this->botToken, $mockHttpClient);

        $apiService->getUserFromGuild($this->guildId, $this->discordId);
    }

    private function generateRandomString($length = 50) {
        return substr(str_shuffle(str_repeat($x='0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ', ceil($length/strlen($x)) )),1,$length);
    }
}


tests/Unit/Service/RoutmouteDiscordOAuthServiceTest.php
<?php

namespace Routmoute\Bundle\RoutmouteDiscordBundle\Tests\Unit\Service;

use PHPUnit\Framework\TestCase;
use Symfony\Component\Security\Csrf\CsrfToken;
use Symfony\Component\HttpClient\MockHttpClient;
use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
use Routmoute\Bundle\RoutmouteDiscordBundle\Service\RoutmouteDiscordOAuthService;

class RoutmouteDiscordOAuthServiceTest extends TestCase
{
    public function testGetRedirectDiscordUrl() {
        $mockHttpClient = new MockHttpClient();

        $csrfTokenMock = $this->getMockBuilder(CsrfToken::class)->disableOriginalConstructor()->getMock();
        $csrfTokenMock
            ->expects($this->once())
            ->method('getValue')
        ;

        $csrfTokenManagerMock = $this->getMockBuilder(CsrfTokenManagerInterface::class)->getMock();
        $csrfTokenManagerMock
            ->expects($this->once())
            ->method('getToken')
            ->willReturn($csrfTokenMock)
        ;

        $client_id = $this->generateRandomString();
        $client_secret = $this->generateRandomString();
        $scope = "identify email";

        $oAuthService = new RoutmouteDiscordOAuthService(
            $client_id,
            $client_secret,
            $scope,
            $mockHttpClient,
            $csrfTokenManagerMock
        );

        $dicordRedirectUri = $this->generateRandomString();

        $this->assertEquals($oAuthService->getRedirectDiscordUrl($dicordRedirectUri), "https://discord.com/api/oauth2/authorize?client_id=" . $client_id . "&redirect_uri=" . urlencode($dicordRedirectUri) . "&response_type=code&scope=" . urlencode($scope));
    }

    private function generateRandomString($length = 50) {
        return substr(str_shuffle(str_repeat($x='0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ', ceil($length/strlen($x)) )),1,$length);
    }
}


src/DependencyInjection/Configuration.php
<?php

namespace Routmoute\Bundle\RoutmouteDiscordBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

class Configuration implements ConfigurationInterface {
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder('routmoute_discord');
        $treeBuilder->getRootNode()
            ->children()
                ->arrayNode('oauth')
                    ->children()
                        ->scalarNode('client_id')
                            ->isRequired()
                            ->cannotBeEmpty()
                        ->end()
                        ->scalarNode('client_secret')
                            ->isRequired()
                            ->cannotBeEmpty()
                        ->end()
                        ->arrayNode('scope')
                            ->isRequired()
                            ->cannotBeEmpty()
                            ->scalarPrototype()->end()
                        ->end()
                    ->end()
                ->end()
                ->arrayNode('api')
                    ->children()
                        ->scalarNode('bot_token')
                            ->isRequired()
                            ->cannotBeEmpty()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;
        return $treeBuilder;
    }
}

src/Service/RoutmouteDiscordApiService.php
<?php

namespace Routmoute\Bundle\RoutmouteDiscordBundle\Service;

use Routmoute\Bundle\RoutmouteDiscordBundle\Exception\DiscordAccessFailedException;
use Symfony\Contracts\HttpClient\HttpClientInterface;
use Symfony\Contracts\HttpClient\ResponseInterface;

class RoutmouteDiscordApiService
{
    private const DISCORD_API = 'https://discord.com/api';

    private $botToken;
    private $httpClient;

    public function __construct(string $bot_token, HttpClientInterface $httpClient)
    {
        $this->botToken = $bot_token;
        $this->httpClient = $httpClient;
    }

    public function getUserFromDiscordId(string $discordId): array
    {
        $apiResponse = $this->sendToDiscordApi('GET', self::DISCORD_API . '/users/' . $discordId);

        $statusCode = $apiResponse->getStatusCode();
        if ($statusCode != 200)
        {
            if ($statusCode == 404) {
                throw new \Exception('User not exist', 404);
            }
            throw new DiscordAccessFailedException();
        }

        return $apiResponse->toArray();
    }

    public function getUserFromGuild(string $guildId, string $discordId): array
    {
        $apiResponse = $this->sendToDiscordApi('GET', self::DISCORD_API . '/guilds/' . $guildId . '/members/' . $discordId);

        $statusCode = $apiResponse->getStatusCode();
        if ($statusCode != 200)
        {
            if ($statusCode == 404) {
                throw new \Exception('User or guild not exist', 404);
            }
            throw new DiscordAccessFailedException();
        }
        
        return $apiResponse->toArray();
    }

    public function getAllUsersFromGuild(string $guildId, int $limit = 1, int $after = 0): array
    {
        $apiResponse = $this->sendToDiscordApi('GET', self::DISCORD_API . '/guilds/' . $guildId . '/members?limit=' . $limit . '&after=' . $after);
        $statusCode = $apiResponse->getStatusCode();
        if ($statusCode != 200)
        {
            if ($statusCode == 404) {
                throw new \Exception('Guild not exist', 404);
            }
            throw new DiscordAccessFailedException();
        }
        return $apiResponse->toArray();
    }

    private function sendToDiscordApi(string $requestType, string $endPoint): ResponseInterface
    {
        return $this->httpClient->request($requestType, $endPoint, [
            'headers' => [
                'Accept' => 'application/json',
                'Authorization' => "Bot ".$this->botToken
            ]
        ]);
    }
}


src/DependencyInjection/RoutmouteDiscordExtension.php
<?php

namespace Routmoute\Bundle\RoutmouteDiscordBundle\DependencyInjection;

use Symfony\Component\Config\FileLocator;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Extension\Extension;
use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;

class RoutmouteDiscordExtension extends Extension {
    public function load(array $configs, ContainerBuilder $container)
    {
        $loader = new YamlFileLoader($container, new FileLocator(__DIR__ . '/../Resources/config'));
        $loader->load('services.yaml');

        $configuration = new Configuration();
        $config = $this->processConfiguration($configuration, $configs);
        
        $scope = '';
        foreach($config['oauth']['scope'] as $value) {
            if ($scope != '') {
                $scope .= ' ';
            }
            $scope .= $value;
        }

        $container->setParameter('routmoute_discord.client_id', $config['oauth']['client_id']);
        $container->setParameter('routmoute_discord.client_secret', $config['oauth']['client_secret']);
        $container->setParameter('routmoute_discord.scope', $scope);
        $container->setParameter('routmoute_discord.bot_token', $config['api']['bot_token']);
    }
}


src/Service/RoutmouteDiscordOAuthService.php
<?php

namespace Routmoute\Bundle\RoutmouteDiscordBundle\Service;

use Routmoute\Bundle\RoutmouteDiscordBundle\Exception\DiscordAccessFailedException;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Security\Csrf\CsrfToken;
use Symfony\Contracts\HttpClient\HttpClientInterface;
use Symfony\Component\Security\Core\Exception\InvalidCsrfTokenException;
use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;

class RoutmouteDiscordOAuthService
{
    private const DISCORD_API = 'https://discord.com/api';
    
    private $clientId;
    private $clientSecret;
    private $scope;
    private $httpClient;
    private $csrfTokenManager;
    
    public function __construct(string $client_id, string $client_secret, string $scope, HttpClientInterface $httpClient, CsrfTokenManagerInterface $csrfTokenManager)
    {
        $this->clientId = $client_id;
        $this->clientSecret = $client_secret;
        $this->scope = $scope;
        $this->httpClient = $httpClient;
        $this->csrfTokenManager = $csrfTokenManager;
    }

    public function getRedirectDiscordUrl(string $redirectUrl): string
    {
        $csrfToken = $this->csrfTokenManager->getToken('routmoute_discord_auth')->getValue();

        $queryParams = http_build_query([
            'client_id' => $this->clientId,
            'redirect_uri' => $redirectUrl,
            'response_type' => 'code',
            'scope' => $this->scope,
            'state' => $csrfToken
        ]);

        return self::DISCORD_API . '/oauth2/authorize?' . $queryParams;
    }

    public function getUserData(Request $request, string $redirectUrl): array
    {
        $state = $request->get('state');

        if ($state && $this->csrfTokenManager->isTokenValid(new CsrfToken('routmoute_discord_auth', $state)))
        {
            $code = $request->get('code');
            if ($code)
            {
                return $this->getUserDataFromAccessToken($this->getAccessTokenFromCode($code, $redirectUrl));
            }
        }

        throw new InvalidCsrfTokenException();
    }
    
    private function getAccessTokenFromCode(string $code, string $redirectUrl): string
    {
        $data = $this->httpClient->request('POST', self::DISCORD_API . '/oauth2/token', [
            'headers' => [
                'Accept' => 'application/json',
                'Content-Type' => 'application/x-www-form-urlencoded'
            ],
            'body' => [
                'client_id' => $this->clientId,
                'client_secret' => $this->clientSecret,
                'code'=> $code,
                'grant_type'=> 'authorization_code',
                'redirect_uri' => $redirectUrl,
                'scope' => $this->scope,
            ]
        ]);

        if ($data->getStatusCode() != 200) {
            throw new DiscordAccessFailedException();
        }

        return $data->toArray()['access_token'];
    }

    private function getUserDataFromAccessToken(string $accessToken): array
    {
        $userData = $this->httpClient->request('GET', self::DISCORD_API . "/users/@me", [
            'headers' => [
                'Accept' => 'application/json',
                'Authorization' => "Bearer {$accessToken}"
            ]
        ]);

        $statusCode = $userData->getStatusCode();
        if ($statusCode != 200)
        {
            if ($statusCode == 404) {
                throw new \Exception('Discord user unattainable');
            }
            throw new DiscordAccessFailedException();
        }

        return $userData->toArray();
    }
}